File: microservice/microservice-app-reactive/src/test/java/com/greg/reactive/CustomerTests.java
package com.greg.reactive;

import com.greg.microsservice.shared.model.CustomerCreateRequest;
import com.greg.microsservice.shared.model.CustomerModel;
import org.junit.jupiter.api.Test;
import org.springframework.http.MediaType;
import reactor.core.publisher.Mono;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

public class CustomerTests extends ControllerTestBase{
    @Test
    public void shouldCreateCustomer() throws Exception {
        webTestClient.post().uri("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .body(Mono.just(new CustomerCreateRequest("bob")),CustomerCreateRequest.class)
                .exchange()
                .expectStatus().isOk();
    }

    @Test
    public void shouldGetAllCustomers() throws Exception {
        var result = webTestClient.get().uri("/customer")
                .exchange()
                .expectStatus().isOk()
                .expectBody()
                .returnResult();

        var customersInitial = asJsonObject(new String(result.getResponseBody()), CustomerModel[].class);

        assertThat(customersInitial.length, equalTo(0));

        webTestClient.post().uri("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .body(Mono.just(new CustomerCreateRequest("bob")),CustomerCreateRequest.class)
                .exchange()
                .expectStatus().isOk();

        webTestClient.post().uri("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .body(Mono.just(new CustomerCreateRequest("bob2")),CustomerCreateRequest.class)
                .exchange()
                .expectStatus().isOk();

        var finalResult = webTestClient.get().uri("/customer")
                .exchange()
                .expectStatus().isOk()
                .expectBody()
                .returnResult();

        var customersFinal = asJsonObject(new String(finalResult.getResponseBody()), CustomerModel[].class);

        assertThat(customersFinal.length, equalTo(2));
    }
}

File: microservice/microservice-app-kotlin/src/test/kotlin/com/greg/microserviceappkotlin/CustomerTests.kt
package com.greg.microserviceappkotlin

import com.greg.microsservice.shared.model.CustomerCreateRequest
import com.greg.microsservice.shared.model.CustomerModel
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

class CustomerTests (@Autowired  mockMvc: MockMvc): ControllerTestBase (mockMvc){
    @Test
    @Throws(Exception::class)
    fun shouldCreateCustomer() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .content(asJsonString(CustomerCreateRequest("Bob")))
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @Throws(Exception::class)
    fun shouldGetAllCustomers() {
        val getCustomersInitial = mockMvc.perform(
            MockMvcRequestBuilders.get("/customer")
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andReturn()
        val customersInitial =
            asJsonObject(getCustomersInitial.response.contentAsString, Array<CustomerModel>::class.java)
        MatcherAssert.assertThat(customersInitial!!.size, Matchers.`is`(0))
        mockMvc.perform(
            MockMvcRequestBuilders.post("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .content(asJsonString(CustomerCreateRequest("Bob")))
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
        mockMvc.perform(
            MockMvcRequestBuilders.post("/customer")
                .contentType(MediaType.APPLICATION_JSON)
                .content(asJsonString(CustomerCreateRequest("Bob")))
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
        val getCustomersFinal = mockMvc.perform(
            MockMvcRequestBuilders.get("/customer")
                .contentType(MediaType.APPLICATION_JSON)
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andReturn()
        val customersFinal =
            asJsonObject(getCustomersFinal.response.contentAsString, Array<CustomerModel>::class.java)
        MatcherAssert.assertThat(customersFinal!!.size, Matchers.`is`(2))
    }
}